sql/2020/03_Markup/pages_almanac_by_device_and_attribute_name_frequency.sql

```diff
@@ -0,0 +1,40 @@
+#standardSQL
+# pages almanac metrics grouped by device and element attribute use (frequency)
+
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+CREATE TEMPORARY FUNCTION get_almanac_attribute_info(almanac_string STRING)
+RETURNS ARRAY<STRUCT<name STRING, freq INT64>> LANGUAGE js AS '''
+try {
+    var almanac = JSON.parse(almanac_string);
+
+    if (Array.isArray(almanac) || typeof almanac != 'object') return [];
+
+    if (almanac.attributes_used_on_elements) {
+      return Object.entries(almanac.attributes_used_on_elements).map(([name, freq]) => ({name, freq})); 
+    }
+    
+} catch (e) {
+    
+}
+return []; 
+''';
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  almanac_attribute_info.name,
+  SUM(almanac_attribute_info.freq) AS freq, # total count from all pages
+  AS_PERCENT(SUM(almanac_attribute_info.freq), SUM(SUM(almanac_attribute_info.freq)) OVER (PARTITION BY _TABLE_SUFFIX)) AS pct_m400 
+FROM
+  `httparchive.pages.2020_08_01_*`,
+  UNNEST(get_almanac_attribute_info(JSON_EXTRACT_SCALAR(payload, '$._almanac'))) AS almanac_attribute_info
```

Comment:
Nit: using `almanac` in the variable/function names is redundant. (Throughout)

Reply:
Major refactor. I specifically named them at first because I was testing getting multiple metrics in the one query. I still like to name them as each accepts a different parameter and it's a reminder of that. Don't pass _markup into an almanac function. If JavaScript was strongly typed and allowed method overloading I would not name them ;-)

I did think about using common names for all queries at one point. I think I backed away as it would be hard to reverse if required. I would be comfortable in going that direction now I know how it all worked out. But I don't want to do a major refactor like that now. The author is nagging me because we've missed a deadline.

Another confusion I saw was that I used _info for the SQL field names, but not in the JavaScript. I picked up quite a few errors from mixing them up because of that.